bin/cheribsdtest/cheribsdtest.c

```diff
@@ -756,6 +756,12 @@ main(int argc, char *argv[])
 			err(EX_OSERR, "setrlimit");
 	}
 
+	/*
```

Comment:
This is after any .init routines have run from DSOs, etc. so seems rather late?  I guess you are just trying to save the PCC of main but not of the function pointer invoked?  I think though this means that main() still has all the SWPERMS enabled, as well as anything main() calls directly in the executable without indirecting via a PLT stub?  That inconsistency seems broken.  Surely main() should run with the same permissions as other function pointers?

Reply:
See above comment about the main() PCC.
Anyway, unrelated to the patch, I think this depends on the definition of initreg. Currently the tests appear to assume that the init registers (and specifically PCC) will not be changed either by .init routines, PLT trampolines or the main() function itself. For testing purposes we should probably define what are initregs and whether there are multiple versions of initregs, because then we may allow .init routines and PLT trampolines that do more fancy things for compartmentalisation or any other purpose.

Reply:
I'm less certain that initregs has as much value as it did when first added as we know trust our toolchain much more than when these tests were probably added.  I do think we want main() and GOT function pointers to probably run with the same set of permissions.

That said, I think your other conclusion about just managing SW_VMEM for now is probably correct.  I do think it's hard to reason about undefined SW bits given the lack of semantics, but I think barring anything else the behavior should be consistent between main() and functions called via GOT pointers.  I think the inconsistency is too confusing otherwise.

Reply:
So the current patch yanks out the undefined SW_PERM clearing and enforces SW_VMEM clearing. I also added a test, which quite simple but should catch regressions there.